File: 2020/day15/day15.py

def game(A, N):
  spoken = {v: k + 1 for k, v in enumerate(A)}
  for i in range(len(A), N):
    prev = spoken.get(A[-1])
    A.append(i - prev if prev is not None else 0)
    spoken[A[-2]] = i
  return A[-1]

S = open('day15.txt').read()
A = list(map(int, S.split(',')))
print(game(A, 2020))
print(game(A, 30000000))

File: 2020/day15/day15.kt
import java.io.File

fun play(input: List<Int>, N: Int): Int {
    val spoken = input.mapIndexed { k, v -> v to k + 1 }
        .toMap().toMutableMap()
    var prev = input.last()
    for (i in input.size..(N - 1)) {
        val pos = spoken.get(prev)
        spoken[prev] = i
        prev = if (pos != null) i - pos else 0
    }
    return prev
}

fun main() {
    val input = File("day15.txt").readText().trim()
        .split(',').map { it.toInt() }
    println(play(input, 2020))
    println(play(input, 30000000))
}
